Alerting: Change screenshots to use components

grobinson-grafana opened the pull request:
**What this PR does / why we need it**:

This commit changes screenshots to use a number of components instead of a set of functional wrappers. It also moves the uploading of screenshots from the screenshot package to the image package so we can re-use the same code for both uploading screenshots and server-side images.

**Which issue(s) this PR fixes**:

Fixes #49332

**Special notes for your reviewer**:

grobinson-grafana committed:
Alerting: Change screenshots to use components

This commit changes screenshots to use a number of components instead
of a set of functional wrappers. It also moves the uploading of screenshots
from the screenshot package to the image package so we can re-use the same
code for both uploading screenshots and server-side images.

joeblubaugh commented on pkg/services/screenshot/singleflight.go:
```diff
@@ -0,0 +1,47 @@
+	f singleflight.Group
+}
+
+func NewSingleFlight() SingleFlight {
```
Do you need this constructor? The concrete type is already exported.

grobinson-grafana replied:
It's not needed for initialization, but I thought it would be nice when constructing to have:

```
s = screenshot.NewManagedScreenshotService(
    screenshot.NewHeadlessCaptureService(ds, rs),
    screenshot.NewSingleFlight(),
    screenshot.NewFixedTokenBucket(cfg.UnifiedAlerting.Screenshots.MaxConcurrentScreenshots),
    screenshot.NewInmemCacheService(screenshotCacheTTL, r),
    r)
```

instead of something like this:

```
s = screenshot.NewManagedScreenshotService(
    screenshot.NewHeadlessCaptureService(ds, rs),
    &screenshot.SingleFlight{},
    screenshot.NewFixedTokenBucket(cfg.UnifiedAlerting.Screenshots.MaxConcurrentScreenshots),
    screenshot.NewInmemCacheService(screenshotCacheTTL, r),
    r)
```

joeblubaugh commented on pkg/services/screenshot/singleflight.go:
```diff
@@ -0,0 +1,47 @@
+func (s *SingleFlighter) Do(ctx context.Context, opts ScreenshotOptions, fn captureFunc) (*Screenshot, error) {
+	k := fmt.Sprintf("%s-%d-%s", opts.DashboardUID, opts.PanelID, opts.Theme)
+
+	v, err, _ := s.f.Do(k, func() (interface{}, error) {
```
Why don't we care about this value? If the result is shared, we might do duplicate work in each caller, right?

grobinson-grafana replied:
You're right, I think we should! For example, we would want to know about it to avoid uploading the same image more than once, and even saving the metadata to the database multiple times.

grobinson-grafana replied:
I made some changes following this feedback!

1. I moved SingleFlight from screenshot package to the image package so we can do a singeflight of not just taking the screenshot, but also uploading it and saving it to the database
2. I removed the use of (but not the code) the cache because I don't think it makes sense to cache the image across evaluations, just across alert instances in the same evaluation

joeblubaugh commented on pkg/services/screenshot/screenshot.go:
```diff
@@ -84,70 +31,28 @@ type ScreenshotService interface {
-		service: service,
+func NewManagedScreenshotService(cs CaptureService, sf SingleFlight, tb TokenBucket, ch CacheService, r prometheus.Registerer) ScreenshotService {
+	return &ManagedScreenshotService{
+		cache:        ch,
```
Do you want to support a "no cache" service? There's a RAM cache and a mock right now. Do we just reduce the expiration time to zero to get "don't cache" functionality? That works, but it allocates a lot of objects before immediately expiring them.

joeblubaugh commented on pkg/services/screenshot/screenshot.go:
```diff
@@ -84,70 +31,28 @@ type ScreenshotService interface {
-func NewObservableScreenshotService(r prometheus.Registerer, service ScreenshotService) ScreenshotService {
-	return &ObservableScreenshotService{
-		service: service,
+func NewManagedScreenshotService(cs CaptureService, sf SingleFlight, tb TokenBucket, ch CacheService, r prometheus.Registerer) ScreenshotService {
```
Is now the time to `Provide` a screenshotservice? Does that buy us anything?

grobinson-grafana replied:
I think not at this time, for the same reason discussed here https://github.com/grafana/grafana/issues/49296#issuecomment-1250231115

joeblubaugh reviewed (approved):
LGTM. A few questions, nothing blocking.

grobinson-grafana committed:
Add NoCacheService

grobinson-grafana committed:
Alerting: Move SingleFlight to image package and remove cache

This commit moves SingleFlight from the screenshot package to the image
package so we can use it for both taking and uploading the screenshot,
where as before it was used just for taking the screenshot.

It also removes the use of a cache because we know that screenshots
can be taken at most once per tick of the scheduler.
